jetty-core/jetty-compression/jetty-compression-zstandard/src/main/java/org/eclipse/jetty/compression/zstandard/ZstandardCompression.java

```diff
@@ -0,0 +1,195 @@
+import com.github.luben.zstd.ZstdOutputStreamNoFinalizer;
+import org.eclipse.jetty.compression.Compression;
+import org.eclipse.jetty.compression.DecoderConfig;
+import org.eclipse.jetty.compression.DecoderSource;
+import org.eclipse.jetty.compression.EncoderConfig;
+import org.eclipse.jetty.compression.EncoderSink;
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.PreEncodedHttpField;
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Compression for Zstandard.
+ *
+ * <p>
+ * Note about {@link ByteBufferPool}: the {@code zstd-jni} project requires {@link java.nio.ByteBuffer}
+ * implementations that are array backed with a zero arrayOffset.
+ * </p>
+ *
+ * @see <a href="https://datatracker.ietf.org/doc/html/rfc8478">RFC 8478 - Zstandard Compression and the application/zstd Media Type</a>
+ * @see <a href="https://github.com/luben/zstd-jni">Uses zstd-jni</a>
+ */
+public class ZstandardCompression extends Compression
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ZstandardCompression.class);
+
+    private static final String ENCODING_NAME = "zstd";
+    private static final HttpField X_CONTENT_ENCODING = new PreEncodedHttpField("X-Content-Encoding", ENCODING_NAME);
+    private static final HttpField CONTENT_ENCODING = new PreEncodedHttpField(HttpHeader.CONTENT_ENCODING, ENCODING_NAME);
+    private static final int DEFAULT_MIN_ZSTD_SIZE = 48;
+    private static final List<String> EXTENSIONS = List.of("zst");
+    private int minCompressSize = DEFAULT_MIN_ZSTD_SIZE;
+    private ZstandardEncoderConfig defaultEncoderConfig = new ZstandardEncoderConfig();
+    private ZstandardDecoderConfig defaultDecoderConfig = new ZstandardDecoderConfig();
+
+    public ZstandardCompression()
+    {
+        super(ENCODING_NAME);
+    }
+
+    @Override
+    public RetainableByteBuffer acquireByteBuffer()
+    {
+        return acquireByteBuffer(getBufferSize());
+    }
+
+    @Override
+    public RetainableByteBuffer acquireByteBuffer(int length)
+    {
+        // Zero-capacity buffers aren't released, they MUST NOT come from the pool.
+        if (length == 0)
+            return RetainableByteBuffer.EMPTY;
+
+        // Per zstd-jni, these MUST be direct ByteBuffer implementations.
+        RetainableByteBuffer.Mutable buffer = getByteBufferPool().acquire(length, true);
+        if (!buffer.getByteBuffer().isDirect())
+        {
+            buffer.release();
+            throw new IllegalStateException("ByteBufferPool does not return zstd-jni required direct ByteBuffer");
+        }
+        // We rely on the ByteBufferPool.release(ByteBuffer) performing a ByteBuffer order reset to default (big-endian).
+        // Typically, this is done with a BufferUtil.reset(ByteBuffer) call on.
```

Comment:
Adding a test to the current pool's suite to ensure byte ordering is properly reset sounds advisable.

Reply:
Since GzipHandler already does this ByteBuffer.order change (after acquire), I don't see this as a big deal.

If we decide to address this with more testing, then we should do it for all of our new ByteBufferPool implementations and also review other places where this order change occurs.

That would make for a good (but separate) PR.

Reply:
Opened a separate (initial) PR for this at https://github.com/jetty/jetty.project/pull/12432